examples/platforms/cc2538/radio.c

```diff
@@ -432,3 +430,265 @@ void RFCoreErrIntHandler(void)
+    }
+    else
+    {
+        ((volatile uint32_t *)RFCORE_XREG_SRCEXTEN0)[index] &= ~((bitMask) << (entry % 8));
+    }
+}
+
+int8_t findEmptySrcMatchAddressEntry(bool aShort)
+{
+    uint32_t bitMask;
+    uint32_t shortEnableStatus = getSrcMatchEnableStatus(true);
+    uint32_t extEnableStatus = getSrcMatchEnableStatus(false);
+    int8_t entry = -1;
+
+    if (aShort)
+    {
+        bitMask = 0x00000001;
+
+        for (uint8_t i = 0; i < 24; i++)
+        {
+            if ((extEnableStatus & bitMask) == 0)
+            {
+                if ((shortEnableStatus & bitMask) == 0)
+                {
+                    ExitNow(entry = i);
+                }
+            }
+
+            if (i % 2 == 1)
+            {
+                extEnableStatus = extEnableStatus >> 2;
+            }
+
+            shortEnableStatus = shortEnableStatus >> 1;
+
+        }
+    }
+    else
+    {
+        bitMask = 0x00000003;
+
+        for (uint8_t i = 0; i < 12; i++)
+        {
+            if (((extEnableStatus | shortEnableStatus) & bitMask) == 0)
+            {
+                ExitNow(entry = i);
+            }
+
+            extEnableStatus = extEnableStatus >> 2;
+            shortEnableStatus = shortEnableStatus >> 2;
+        }
+    }
+
+exit:
+    return entry;
+}
+
+ThreadError otPlatRadioAddShortSrcMatchAddressEntry(uint16_t aShortAddress, uint16_t aPanId)
+{
+    ThreadError error = kThreadError_None;
+    int8_t entry = findEmptySrcMatchAddressEntry(true);
+
+    if (entry < 0)
+    {
+        ackPendingAll();
+    }
+    else
+    {
+        uint32_t addr = RFCORE_FFSM_SRCADDRESS_TABLE + entry * 16;
+        ((volatile uint32_t *)addr)[0] = aPanId & 0xFF;
+        ((volatile uint32_t *)addr)[1] = aPanId >> 8;
+        ((volatile uint32_t *)addr)[2] = aShortAddress & 0xFF;
+        ((volatile uint32_t *)addr)[3] = aShortAddress >> 8;
+        setSrcMatchAddressEnable(true, (uint8_t)(entry), true);
+    }
+
+    return error;
+}
+
+ThreadError otPlatRadioAddExtSrcMatchAddressEntry(uint8_t *aExtAddress)
+{
+    ThreadError error = kThreadError_None;
+    int8_t entry = findEmptySrcMatchAddressEntry(false);
+
+    if (entry < 0)
+    {
+        ackPendingAll();
+    }
+    else
+    {
+        uint32_t addr = RFCORE_FFSM_SRCADDRESS_TABLE + entry * 32;
+
+        for (uint8_t i = 0; i < 8; i++)
+        {
+            ((volatile uint32_t *)addr)[i] = aExtAddress[i];
+        }
+
+        setSrcMatchAddressEnable(false, (uint8_t)(entry), true);
+    }
+
+    return error;
+}
+
+ThreadError otPlatRadioClearShortSrcMatchAddressEntry(uint16_t aShortAddress, uint16_t aPanId)
+{
+    ThreadError error = kThreadError_None;
+    int8_t entry = findShortSrcMatchAddressEntry(aShortAddress, aPanId);
+    VerifyOrExit(entry >= 0, error = kThreadError_NoAddress);
+
+    setSrcMatchAddressEnable(true, (uint8_t)(entry), false);
+
+    if (isAckPendingAll())
+    {
+        ackPendingPoll();
```

Comment:
I think we need to think carefully about the case when the table overflows.  What if we've overflowed the hardware table by more than one entry?  We can either:

1. Reference count the number of attempts `otPlatRadioAdd*` was called and no entries were available.
2. Have the upper layer flush the table and re-add entries again.